### PR TITLE
feat(completion):  completeopt support fuzzy

### PR DESCRIPTION
**Pull request by glepnir**
Add fuzzy into completeopt

context #14694

### CODE REVIEW

File: src/insexpand.c [hunk 1]
```diff
@@ -1212,6 +1219,17 @@ trigger_complete_changed_event(int cur)
...
 
+/*
+ * pumitem qsort compare func
+ */
+    static int
+ins_compl_fuzzy_sort(const void *a, const void *b)
+{
+  const int sa = (*(pumitem_T *)a).pum_score;
+  const int sb = (*(pumitem_T *)b).pum_score;
+  return sa == sb ? 0 : sa < sb ? 1 : -1;
```
Comment (yegappan) [inline src/insexpand.c]: Is proper indentation (four spaces) used in this function?

File: src/insexpand.c [hunk 2]
```diff
@@ -1236,9 +1255,17 @@ ins_compl_build_pum(void)
 
     do
     {
+	// when completeopt include fuzzy option and leader is not null or empty
+	// set the cp_score for after compare.
+	if (compl_fuzzy_match == TRUE && compl_leader != NULL && lead_len > 0)
+	{
```
Comment (yegappan) [inline src/insexpand.c]: The coding convention is to not use curly braces if there is a single statement in the body of a if statement.

File: src/insexpand.c [hunk 3]
```diff
@@ -1236,9 +1255,17 @@ ins_compl_build_pum(void)
 
     do
     {
+	// when completeopt include fuzzy option and leader is not null or empty
+	// set the cp_score for after compare.
+	if (compl_fuzzy_match == TRUE && compl_leader != NULL && lead_len > 0)
```
Comment (yegappan) [inline src/insexpand.c]: As "compl_fuzzy_match" is a boolean variable, there is no need to compare it against "TRUE".

File: src/insexpand.c [hunk 4]
```diff
@@ -1236,9 +1255,17 @@ ins_compl_build_pum(void)
...
+	{
+	    compl->cp_score = fuzzy_match_str(compl->cp_str, compl_leader);
+	}
+
 	if (!match_at_original_text(compl)
 		&& (compl_leader == NULL
-		    || ins_compl_equal(compl, compl_leader, lead_len)))
+		    || ins_compl_equal(compl, compl_leader, lead_len)
+		    || (compl_fuzzy_match == TRUE && compl->cp_score > 0)))
```
Comment (yegappan) [inline src/insexpand.c]: No need to compare "compl_fuzzy_match" against TRUE.

File: src/insexpand.c [hunk 5]
```diff
@@ -1267,9 +1294,10 @@ ins_compl_build_pum(void)
     {
 	if (!match_at_original_text(compl)
 		&& (compl_leader == NULL
-		    || ins_compl_equal(compl, compl_leader, lead_len)))
+		    || ins_compl_equal(compl, compl_leader, lead_len)
+		    || (compl_fuzzy_match == TRUE && compl->cp_score > 0)))
```
Comment (yegappan) [inline src/insexpand.c]: Remove comparison to TRUE.

File: src/insexpand.c [hunk 6]
```diff
@@ -1267,9 +1294,10 @@ ins_compl_build_pum(void)
...
 	{
-	    if (!shown_match_ok)
+	    if (!shown_match_ok && compl_fuzzy_match == FALSE)
```
Comment (yegappan) [inline src/insexpand.c]: Can you use "!compl_fuzzy_match" instead of comparing against FALSE?

File: src/insexpand.c [hunk 7]
```diff
@@ -4025,6 +4081,24 @@ ins_compl_show_filename(void)
     redraw_cmdline = FALSE;	    // don't overwrite!
 }
 
+    static compl_T *
+find_next_comp_when_fuzzy(void)
```
Comment (yegappan) [inline src/insexpand.c]: Can you add a test for this function?

**Review by yegappan — CHANGES_REQUESTED** [review]

**Comment by glepnir** [comment]
I tried it briefly. but not sure yet because I don't fully understand some parts of the code ..

**Comment by glepnir** [comment]
@chrisbra PTAL

File: src/insexpand.c [hunk 8]
```diff
@@ -207,6 +209,8 @@ static int	  compl_cont_status = 0;
 static int	  compl_opt_refresh_always = FALSE;
 static int	  compl_opt_suppress_empty = FALSE;
 
+static int pum_selected_item = -1;
```
Comment (chrisbra) [inline src/insexpand.c]: shouldn't that be rather `compl_selected_item` to align with the other compl items? Also, please align with the other static ints above.

File: src/testdir/test_ins_complete.vim [hunk 9]
```diff
@@ -2451,4 +2451,56 @@ func Test_completefunc_first_call_complete_add()
   bwipe!
 endfunc
 
+func Test_complete_fuzzy_match()
+  function! OnPumChange()
+    let g:event = copy(v:event)
```
Comment (chrisbra) [inline src/testdir/test_ins_complete.vim]: this seems unused?

**Comment by chrisbra** [comment]
Thanks, I think it looks reasonable. Can you please also add some documentation at `:h 'completeopt'`?

File: src/insexpand.c (hunk 8 above)
Comment (glepnir) [reply src/insexpand.c]: because this variable take from neovim  :)

File: src/testdir/test_ins_complete.vim [hunk 10]
```diff
@@ -2451,4 +2451,55 @@ func Test_completefunc_first_call_complete_add()
   bwipe!
 endfunc
 
+func Test_complete_fuzzy_match()
```
Comment (zeertzjq) [inline src/testdir/test_ins_complete.vim:2454]:
```suggestion
func Test_complete_fuzzy_match()
  new

```

File: src/insexpand.c [hunk 11]
```diff
@@ -4025,6 +4077,40 @@ ins_compl_show_filename(void)
...
+    if (compl_match_array == NULL ||
+	    (is_forward && compl_selected_item == compl_match_arraysize - 1)
+	    || (is_backward && compl_selected_item == 0))
+	return compl_first_match;
+
+    if (is_forward)
+	target_idx = compl_selected_item + 1;
+    else if (is_backward)
+      target_idx = compl_selected_item == -1 ? compl_match_arraysize - 1
+					    : compl_selected_item - 1;
```
Comment (yegappan) [inline src/insexpand.c:4099]: Can you check the indentation for the lines in this function?  Vim source base uses four spaces for `shiftwidth`.

File: src/insexpand.c [hunk 12]
```diff
@@ -1320,6 +1365,12 @@ ins_compl_build_pum(void)
 	compl = compl->cp_next;
     } while (compl != NULL && !is_first_match(compl));
 
+    if (compl_fuzzy_match && compl_leader != NULL && lead_len > 0)
+    {
+	// sort by the largest score of fuzzy match
+	qsort(compl_match_array, (size_t)compl_match_arraysize, sizeof(pumitem_T), ins_compl_fuzzy_sort);
+    }
```
Comment (yegappan) [inline src/insexpand.c:1372]: Can you remove the curly braces for this single statement if block?

File: src/insexpand.c [hunk 13]
```diff
@@ -4025,6 +4077,40 @@ ins_compl_show_filename(void)
     redraw_cmdline = FALSE;	    // don't overwrite!
 }
 
+    static compl_T *
```
Comment (yegappan) [inline src/insexpand.c:4080]: Can you add a comment above describing the functionality in this function?

File: src/insexpand.c [hunk 14]
```diff
@@ -1285,6 +1311,24 @@ ins_compl_build_pum(void)
 		    shown_compl = compl;
 		cur = i;
 	    }
+	    else if (compl_fuzzy_match)
+	    {
+		if (compl->cp_score > max_fuzzy_score)
```
Comment (yegappan) [inline src/insexpand.c:1316]: Can you add a comment here describing the functionality in this block?

**Review by yegappan — CHANGES_REQUESTED** [review]

**Comment by chrisbra** [comment]
Opps. sorry, I didn't see Yegappans comments here when I started mergin down this. @glepnir can you make the suggested changes in a follow-up patch please?

File: src/testdir/test_ins_complete.vim (hunk 10 above)
Comment (chrisbra) [reply src/testdir/test_ins_complete.vim:2454]: I fixed this locally when merging

File: src/insexpand.c (hunk 12 above)
Comment (chrisbra) [reply src/insexpand.c:1372]: I'll fixed that already while merging

**Comment by glepnir** [comment]
will do

**Pull request closed without merging**